codex/blockexchange/network/network.nim

```diff
@@ -93,18 +93,20 @@ proc send*(b: BlockExcNetwork, id: PeerId, msg: pb.Message) {.async.} =
   ## Send message to peer
   ##
 
-  b.peers.withValue(id, peer):
-    try:
-      await b.inflightSema.acquire()
-      await peer[].send(msg)
-    except CancelledError as error:
-      raise error
-    except CatchableError as err:
-      error "Error sending message", peer = id, msg = err.msg
-    finally:
-      b.inflightSema.release()
-  do:
+  if not (id in b.peers):
     trace "Unable to send, peer not found", peerId = id
+    return
+
+  let peer = b.peers[id]
```

Comment:
Interesting, it makes sense that the object can get collected in between `acquire` and `send`. It would have been probably better if `withValue` would also increase the refcount with `GC_Ref/Unref` for reference types, so that this situations don't occur.

That said, I think the real culprit is the semaphore, I cannot remember exactly why I had it in there in the first place and it's only used here, but I think I wanted to avoid this type of situations by synchronizing access to to the peers table, but that isn't a great idea either way.

I would remove the semaphore entirely as it isn't doing much here either way and peers should be able to handle concurrent sends.

Reply:
I'm of two minds. I want the semaphore gone. I built an image without it and throw a bunch of tests at it. It seems to work all right. But the issue you found and linked has me worried.

So I propose we merge the fix by replacing the .withValue. And we make sure we never use withValue in an async context. (I checked all other uses of withValue and none of them are at risk!)

We can make a separate issue for the semaphore. I suspect a semaphore per peer might be a sensible approach: it allows more to go on simultaneously but still limits the outbound requests.

Reply:
Might be a good idea to add a comment on `withValue` that instructs our future selves not to use it in within an async context, referencing this PR and https://github.com/codex-storage/nim-codex/issues/914.

Reply:
This. `withValue` needs manual refcount tracking with async due to raw pointers. Imo, I don't think the semaphore is needed in this context.

Reply:
Created an issue for this: https://github.com/codex-storage/nim-codex/issues/943